src/services/identity/CollaboratorsService.ts

```diff
@@ -0,0 +1,239 @@
+import _ from "lodash"
+import { ModelStatic, Op } from "sequelize"
+import validator from "validator"
+
+import { ForbiddenError } from "@errors/ForbiddenError"
+import { NotFoundError } from "@errors/NotFoundError"
+import { UnprocessableError } from "@errors/UnprocessableError"
+
+import { CollaboratorRoles } from "@constants/constants"
+
+import { Whitelist, User, Site, SiteMember } from "@database/models"
+import { BadRequestError } from "@root/errors/BadRequestError"
+import { ConflictError } from "@root/errors/ConflictError"
+import logger from "@root/logger/logger"
+
+import SitesService from "./SitesService"
+import UsersService from "./UsersService"
+
+interface CollaboratorsServiceProps {
+  siteRepository: ModelStatic<Site>
+  siteMemberRepository: ModelStatic<SiteMember>
+  sitesService: SitesService
+  usersService: UsersService
+  whitelist: ModelStatic<Whitelist>
+}
+
+class CollaboratorsService {
+  // NOTE: Explicitly specifying using keyed properties to ensure
+  // that the types are synced.
+  private readonly siteRepository: CollaboratorsServiceProps["siteRepository"]
+
+  private readonly siteMemberRepository: CollaboratorsServiceProps["siteMemberRepository"]
+
+  private readonly sitesService: CollaboratorsServiceProps["sitesService"]
+
+  private readonly usersService: CollaboratorsServiceProps["usersService"]
+
+  private readonly whitelist: CollaboratorsServiceProps["whitelist"]
+
+  constructor({
+    siteRepository,
+    siteMemberRepository,
+    sitesService,
+    usersService,
+    whitelist,
+  }: CollaboratorsServiceProps) {
+    this.siteRepository = siteRepository
+    this.siteMemberRepository = siteMemberRepository
+    this.sitesService = sitesService
+    this.usersService = usersService
+    this.whitelist = whitelist
+  }
+
+  deriveAllowedRoleFromEmail = async (fullEmail: string) => {
+    const whitelistEntries = await this.whitelist.findAll({
+      where: {
+        expiry: {
+          [Op.or]: [{ [Op.is]: null }, { [Op.gt]: new Date() }],
+        },
+      },
+    })
+
+    const matchedDomains = whitelistEntries.filter((entry) =>
+      fullEmail.endsWith(entry.email)
+    )
+
+    if (!matchedDomains.length) return null
+
+    // TODO: Modify this method because the presence of the expiry field is not
+    // the best way of differentiating Admin/Contributor roles
+    return matchedDomains[0].expiry
+      ? CollaboratorRoles.Contributor
+      : CollaboratorRoles.Admin
```

Comment:
I wanted to keep the scope of this PR limited to just `Contributor` and `Admin` because we discussed just these two roles with the design team - we could add the `SuperAdmin` features later in a separate PR